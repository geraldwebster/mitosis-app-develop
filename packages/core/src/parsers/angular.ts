import ts from 'typescript';
import { parseTemplate } from '@angular/compiler';
import {
  Node,
  Element,
  Template,
  Text,
  BoundText,
} from '@angular/compiler/src/render3/r3_ast';
import { ASTWithSource } from '@angular/compiler/src/expression_parser/ast';
import { createJSXLiteComponent } from '../helpers/create-jsx-lite-component';
import { createJSXLiteNode } from '../helpers/create-jsx-lite-node';
import { JSXLiteNode } from '../types/jsx-lite-node';
import { omit } from 'lodash';
import { babelTransformCode } from '../helpers/babel-transform';
import { types } from '@babel/core';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { capitalize } from '../helpers/capitalize';

const getTsAST = (code: string) => {
  return ts.createSourceFile('code.ts', code, ts.ScriptTarget.Latest, true);
};

interface AngularToJsxLiteOptions {}

const transformBinding = (
  binding: string,
  _options: AngularToJsxLiteOptions,
) => {
  return babelTransformCode(binding, {
    Identifier(path: babel.NodePath<babel.types.Identifier>) {
      const name = path.node.name;

      if (
        (types.isObjectProperty(path.parent) &&
          path.parent.key === path.node) ||
        (types.isMemberExpression(path.parent) &&
          path.parent.property === path.node)
      ) {
        return;
      }

      if (
        !(name.startsWith('state.') || name === 'event' || name === '$event')
      ) {
        path.replaceWith(types.identifier(`state.${name}`));
      }
    },
  });
};

const isElement = (node: Node): node is Element =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).attributes);

const isTemplate = (node: Node): node is Template =>
  // TODO: theres got to be a better way than this
  Array.isArray((node as any).templateAttrs);

const isText = (node: Node): node is Text =>
  typeof (node as any).value === 'string';

const isBoundText = (node: Node): node is BoundText =>
  typeof (node as any).value === 'object';

const angularTemplateNodeToJsxLiteNode = (
  node: Node,
  options: AngularToJsxLiteOptions,
): JSXLiteNode => {
  if (isTemplate(node)) {
    const ngIf = node.templateAttrs.find((item) => item.name === 'ngIf');
    if (ngIf) {
      return createJSXLiteNode({
        name: 'Show',
        bindings: {
          when: transformBinding(
            (ngIf.value as ASTWithSource).source!,
            options,
          ),
        },
        children: [
          angularTemplateNodeToJsxLiteNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
    const ngFor = node.templateAttrs.find((item) => item.name === 'ngFor');
    if (ngFor) {
      const value = (ngFor.value as ASTWithSource).source!;
      const split = value.split(/let\s|\sof\s/);
      const [_let, itemName, _of, expression] = split;
      return createJSXLiteNode({
        name: 'For',
        bindings: {
          each: transformBinding(expression, options),
          _forName: itemName,
        },
        children: [
          angularTemplateNodeToJsxLiteNode(
            omit(node, 'templateAttrs'),
            options,
          ),
        ],
      });
    }
  }

  if (isElement(node)) {
    const properties: Record<string, string> = {};
    const bindings: Record<string, string> = {};

    for (const input of node.inputs) {
      bindings[input.name] = transformBinding(
        (input.value as ASTWithSource).source!,
        options,
      );
    }
    for (const output of node.outputs) {
      bindings['on' + capitalize(output.name)] = transformBinding(
        (output.handler as ASTWithSource)
          .source! // TODO: proper reference replace
          .replace(/\$event/g, 'event'),
        options,
      );
    }
    for (const attribute of node.attributes) {
      properties[attribute.name] = attribute.value;
    }

    return createJSXLiteNode({
      name: node.name,
      properties,
      bindings,
      children: node.children.map((node) =>
        angularTemplateNodeToJsxLiteNode(node, options),
      ),
    });
  }

  if (isText(node)) {
    return createJSXLiteNode({
      properties: {
        _text: node.value,
      },
    });
  }

  if (isBoundText(node)) {
    // TODO: handle the bindings
    return createJSXLiteNode({
      properties: {
        _text: (node.value as ASTWithSource).source!,
      },
    });
  }

  throw new Error(`Element node type {${node}} is not supported`);
};

const angularTemplateToJsxLiteNodes = (
  template: string,
  options: AngularToJsxLiteOptions,
) => {
  const ast = parseTemplate(template, '.');
  const blocks = ast.nodes.map((node) =>
    angularTemplateNodeToJsxLiteNode(node, options),
  );

  return blocks;
};

const parseTypescript = (code: string, options: AngularToJsxLiteOptions) => {
  const component = createJSXLiteComponent();

  const ast = getTsAST(code);
  for (const statement of ast.statements) {
    if (ts.isClassDeclaration(statement)) {
      if (statement.decorators) {
        for (const decorator of statement.decorators) {
          // TODO: proper reference tracing
          if (ts.isCallExpression(decorator.expression))
            if (
              ts.isIdentifier(decorator.expression.expression) &&
              decorator.expression.expression.text === 'Component'
            ) {
              const firstArg = decorator.expression.arguments[0];
              if (ts.isObjectLiteralExpression(firstArg)) {
                firstArg.properties.find((item) => {
                  if (ts.isPropertyAssignment(item)) {
                    if (
                      ts.isIdentifier(item.name) &&
                      item.name.text === 'template'
                    ) {
                      if (ts.isTemplateLiteral(item.initializer)) {
                        const template = item.initializer
                          .getText()
                          .trim()
                          .slice(1, -1);
                        component.children = angularTemplateToJsxLiteNodes(
                          template,
                          options,
                        );
                      }
                    }
                  }
                });
              }
            }
        }
      }
    }
  }

  return component;
};

export function angularToJsxLiteComponent(
  code: string,
  options: AngularToJsxLiteOptions = {},
): JSXLiteComponent {
  return parseTypescript(code, options);
}
