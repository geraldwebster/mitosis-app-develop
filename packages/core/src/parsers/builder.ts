import { BuilderContent, BuilderElement } from '@builder.io/sdk';
import json5 from 'json5';
import { last, omit, pickBy } from 'lodash';
import { createJSXLiteComponent } from '../helpers/create-jsx-lite-component';
import { createJSXLiteNode } from '../helpers/create-jsx-lite-node';
import { JSXLiteNode } from '../types/jsx-lite-node';

// Omit some superflous styles that can come from Builder's web importer
const styleOmitList: (
  | keyof CSSStyleDeclaration
  | 'backgroundRepeatX'
  | 'backgroundRepeatY'
)[] = [
  'backgroundRepeatX',
  'backgroundRepeatY',
  'backgroundPositionX',
  'backgroundPositionY',
  'borderBottomLeftRadius',
  'borderBottomRightRadius',
  'borderTopRightRadius',
  'borderTopRightRadius',
];

const hasStyles = (block: BuilderElement) => {
  if (block.responsiveStyles) {
    for (const key in block.responsiveStyles) {
      if (Object.keys((block.responsiveStyles as any)[key]!).length) {
        return true;
      }
    }
  }
  return false;
};

type InternalOptions = {
  skipMapper?: boolean;
};

const componentMappers: {
  [key: string]: (
    block: BuilderElement,
    options: BuilerToJSXLiteOptions,
  ) => JSXLiteNode;
} = {
  Symbol(block, options) {
    const node = builderElementToJsxLiteNode(
      omit(block, 'component.options.symbol.content'),
      options,
      {
        skipMapper: true,
      },
    );

    // TODO: full component code in a new component and hoist it. will need to pass through a `context` object, maybe on options
    const blocks = block.component?.options?.symbol?.content?.data?.blocks;
    if (blocks) {
      node.children = blocks.map((child: any) =>
        builderElementToJsxLiteNode(child, options),
      );
      node.bindings.useChildren = 'true';
    }

    return node;
  },
  Columns(block, options) {
    const node = builderElementToJsxLiteNode(block, options, {
      skipMapper: true,
    });

    delete node.bindings.columns;
    delete node.properties.columns;

    node.children = block.component?.options.columns.map((col: any) =>
      createJSXLiteNode({
        name: 'Column',
        children: col.blocks.map((col: any) =>
          builderElementToJsxLiteNode(col, options),
        ),
      }),
    );

    return node;
  },
  'Shopify:For': (block, options) => {
    return createJSXLiteNode({
      name: 'For',
      bindings: {
        _forName: block.component!.options!.repeat!.itemName,
        each: `state.${block.component!.options!.repeat!.collection}`,
      },
      children: (block.children || []).map((child) =>
        builderElementToJsxLiteNode(child, options),
      ),
    });
  },
  Text: (block, options) => {
    const bindings: any = {
      ...omit(block.bindings, 'component.options.text'),
      ...(Boolean(Object.keys(block.responsiveStyles?.large || {}).length) && {
        css: JSON.stringify(block.responsiveStyles?.large || {}),
      }),
    };
    const properties = { ...block.properties };

    const innerBindings = {
      _text: block.bindings?.['component.options.text'],
    };
    const innerProperties = {
      _text: block.component!.options.text,
    };

    if ((block.tagName && block.tagName !== 'div') || hasStyles(block)) {
      return createJSXLiteNode({
        bindings,
        properties,
        children: [
          createJSXLiteNode({
            bindings: innerBindings,
            properties: innerProperties,
          }),
        ],
      });
    }
    return createJSXLiteNode({
      name: block.tagName || 'div',
      properties: {
        ...properties,
        ...innerProperties,
      },
      bindings: {
        ...bindings,
        ...innerBindings,
      },
    });
  },
};

export type BuilerToJSXLiteOptions = {
  context?: { [key: string]: any };
  includeBuilderExtras?: boolean;
};
export type InternalBuilerToJSXLiteOptions = BuilerToJSXLiteOptions & {
  context: { [key: string]: any };
};

export const builderElementToJsxLiteNode = (
  block: BuilderElement,
  options: BuilerToJSXLiteOptions = {},
  _internalOptions: InternalOptions = {},
): JSXLiteNode => {
  const mapper =
    !_internalOptions.skipMapper &&
    block.component &&
    componentMappers[block.component!.name];

  if (mapper) {
    return mapper(block, options);
  }

  const bindings: any = {};

  if (block.bindings) {
    for (const key in block.bindings) {
      if (key === 'css') {
        continue;
      }
      const useKey = key.replace(/^(component\.)?options\./, '');
      if (!useKey.includes('.')) {
        bindings[useKey] = block.bindings[key];
      }
    }
  }

  const properties: { [key: string]: string } = {
    ...block.properties,
    ...(options.includeBuilderExtras && {
      ['builder-id']: block.id!,
      class: `builder-block ${block.id} ${block.properties?.class || ''}`,
    }),
  };

  if (block.component?.options) {
    for (const key in block.component.options) {
      const value = block.component.options[key];
      if (typeof value === 'string') {
        properties[key] = value;
      } else {
        bindings[key] = json5.stringify(value);
      }
    }
  }

  if (block.component && block.tagName && block.tagName !== 'div') {
    properties.builderTag = block.tagName;
  }

  return createJSXLiteNode({
    name:
      block.component?.name?.replace(/[^a-z0-9]/gi, '') ||
      block.tagName ||
      'div',
    properties,
    bindings: {
      ...bindings,
      ...(Boolean(Object.keys(block.responsiveStyles?.large || {}).length) && {
        css: JSON.stringify(
          omit(block.responsiveStyles?.large || {}, styleOmitList),
        ),
      }),
    },
    children: (block.children || []).map((item) =>
      builderElementToJsxLiteNode(item, options),
    ),
  });
};

export const builderContentToJsxLiteComponent = (
  builderContent: BuilderContent,
  options: BuilerToJSXLiteOptions = {},
) => {
  const generatedStateMatch = (builderContent?.data?.jsCode || '')
    .trim()
    .match(/Object\.assign\(state\s*,\s*(\{[\s\S]+\})\s*\)/i);

  let state = {};
  if (generatedStateMatch?.[1]) {
    try {
      state = json5.parse(generatedStateMatch[1]);
    } catch (err) {
      console.warn('Error parsing state');
    }
  }

  return createJSXLiteComponent({
    state: {
      ...state,
      ...builderContent.data?.state,
    },
    children: (builderContent.data?.blocks || [])
      .filter((item) => {
        if (item.properties?.src?.includes('/api/v1/pixel')) {
          return false;
        }
        return true;
      })
      .map((item) => builderElementToJsxLiteNode(item, options)),
  });
};
