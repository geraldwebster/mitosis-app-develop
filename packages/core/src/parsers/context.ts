import * as babel from '@babel/core';
import { JSXLiteContext } from '../types/mitosis-context';
import { createJsxLiteContext } from '../helpers/create-mitosis-context';
import { parseStateObject } from './jsx';

const { types } = babel;

const tsPreset = require('@babel/preset-typescript');

type ParseContextOptions = {
  name: string;
};

export function parseContext(
  code: string,
  options: ParseContextOptions,
): JSXLiteContext | null {
  let found = false;
  const context = createJsxLiteContext({ name: options.name });

  babel.transform(code, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [
      () => ({
        visitor: {
          Program(path: babel.NodePath<babel.types.Program>) {
            for (const item of path.node.body) {
              if (types.isExportDefaultDeclaration(item)) {
                const expression = item.declaration;
                if (types.isCallExpression(expression)) {
                  if (
                    types.isIdentifier(expression.callee) &&
                    expression.callee.name === 'createContext'
                  ) {
                    const firstArg = expression.arguments[0];
                    if (types.isObjectExpression(firstArg)) {
                      // TODO: support non object values by parsing any node type
                      // like the logic within each property value of parseStateObject
                      context.value = parseStateObject(firstArg);
                      found = true;
                    }
                  }
                }
              }
            }
          },
        },
      }),
    ],
  });

  if (!found) {
    return null;
  }
  return context;
}
