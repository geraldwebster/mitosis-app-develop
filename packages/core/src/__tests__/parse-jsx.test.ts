import { parseJsx } from '../parsers/jsx';

const buttonWithMetadata = require('./data/blocks/button-with-metadata.raw');

describe('Parse JSX', () => {
  test('metadata', () => {
    const json = parseJsx(buttonWithMetadata);
    expect(json).toMatchSnapshot();
  });
});
