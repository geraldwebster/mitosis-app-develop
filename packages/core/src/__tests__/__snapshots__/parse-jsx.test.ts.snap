// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse JSX metadata 1`] = `
Object {
  "@type": "@jsx-lite/component",
  "children": Array [
    Object {
      "@type": "@jsx-lite/node",
      "bindings": Object {},
      "children": Array [
        Object {
          "@type": "@jsx-lite/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
      ",
          },
        },
        Object {
          "@type": "@jsx-lite/node",
          "bindings": Object {
            "when": "props.link",
          },
          "children": Array [
            Object {
              "@type": "@jsx-lite/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
        ",
              },
            },
            Object {
              "@type": "@jsx-lite/node",
              "bindings": Object {
                "_spread": "props.attributes",
                "href": "props.link",
                "target": "props.openLinkInNewTab ? '_blank' : undefined",
              },
              "children": Array [
                Object {
                  "@type": "@jsx-lite/node",
                  "bindings": Object {},
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {
                    "_text": "
          ",
                  },
                },
                Object {
                  "@type": "@jsx-lite/node",
                  "bindings": Object {
                    "_text": "props.text",
                  },
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {},
                },
                Object {
                  "@type": "@jsx-lite/node",
                  "bindings": Object {},
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {
                    "_text": "
        ",
                  },
                },
              ],
              "meta": Object {},
              "name": "a",
              "properties": Object {},
            },
            Object {
              "@type": "@jsx-lite/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
      ",
              },
            },
          ],
          "meta": Object {},
          "name": "Show",
          "properties": Object {},
        },
        Object {
          "@type": "@jsx-lite/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
      ",
          },
        },
        Object {
          "@type": "@jsx-lite/node",
          "bindings": Object {
            "when": "!props.link",
          },
          "children": Array [
            Object {
              "@type": "@jsx-lite/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
        ",
              },
            },
            Object {
              "@type": "@jsx-lite/node",
              "bindings": Object {
                "_spread": "props.attributes",
              },
              "children": Array [
                Object {
                  "@type": "@jsx-lite/node",
                  "bindings": Object {},
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {
                    "_text": "
          ",
                  },
                },
                Object {
                  "@type": "@jsx-lite/node",
                  "bindings": Object {
                    "_text": "props.text",
                  },
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {},
                },
                Object {
                  "@type": "@jsx-lite/node",
                  "bindings": Object {},
                  "children": Array [],
                  "meta": Object {},
                  "name": "div",
                  "properties": Object {
                    "_text": "
        ",
                  },
                },
              ],
              "meta": Object {},
              "name": "button",
              "properties": Object {
                "type": "button",
              },
            },
            Object {
              "@type": "@jsx-lite/node",
              "bindings": Object {},
              "children": Array [],
              "meta": Object {},
              "name": "div",
              "properties": Object {
                "_text": "
      ",
              },
            },
          ],
          "meta": Object {},
          "name": "Show",
          "properties": Object {},
        },
        Object {
          "@type": "@jsx-lite/node",
          "bindings": Object {},
          "children": Array [],
          "meta": Object {},
          "name": "div",
          "properties": Object {
            "_text": "
    ",
          },
        },
      ],
      "meta": Object {},
      "name": "Fragment",
      "properties": Object {},
    },
  ],
  "hooks": Object {},
  "imports": Array [],
  "meta": Object {
    "useMetadata": Object {
      "foo": Object {
        "bar": "baz",
      },
    },
  },
  "name": "Button",
  "state": Object {},
  "subComponents": Array [],
}
`;
